crates/services/p2p/src/cached_view.rs

```diff
@@ -0,0 +1,320 @@
+            V::get_transactions,
+        )
+    }
+}
+
+#[allow(non_snake_case)]
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use fuel_core_types::blockchain::consensus::Genesis;
+    use std::sync::Arc;
+    use tokio::sync::Notify;
+
+    struct FakeDb {
+        sender: Arc<Notify>,
+        values: bool,
+    }
+
+    #[inline]
+    fn default_sealed_headers(range: Range<u32>) -> Vec<SealedBlockHeader> {
+        vec![SealedBlockHeader::default(); range.len()]
+    }
+
+    #[inline]
+    fn default_transactions(range: Range<u32>) -> Vec<Transactions> {
+        vec![Transactions::default(); range.len()]
+    }
+
+    impl P2pDb for FakeDb {
+        fn get_sealed_headers(
+            &self,
+            range: Range<u32>,
+        ) -> StorageResult<Option<Vec<SealedBlockHeader>>> {
+            self.sender.notify_waiters();
+            if !self.values {
+                return Ok(None);
+            }
+            let headers = default_sealed_headers(range);
+            Ok(Some(headers))
+        }
+
+        fn get_transactions(
+            &self,
+            range: Range<u32>,
+        ) -> StorageResult<Option<Vec<Transactions>>> {
+            self.sender.notify_waiters();
+            if !self.values {
+                return Ok(None);
+            }
+            let transactions = default_transactions(range);
+            Ok(Some(transactions))
+        }
+
+        fn get_genesis(&self) -> StorageResult<Genesis> {
+            self.sender.notify_waiters();
+            Ok(Genesis::default())
+        }
+    }
+
+    #[tokio::test]
+    async fn cached_view__get_sealed_headers__cache_hit() {
+        let sender = Arc::new(Notify::new());
+        let db = FakeDb {
+            sender: sender.clone(),
+            values: true,
+        };
+        let cached_view = CachedView::new(false);
+
+        let block_height_range = 0..100;
+        let sealed_headers = default_sealed_headers(block_height_range.clone());
+        for (block_height, header) in
+            block_height_range.clone().zip(sealed_headers.iter())
+        {
+            cached_view
+                .sealed_block_headers
+                .insert(block_height, header.clone());
+        }
+
+        let result = cached_view
+            .get_sealed_headers(&db, block_height_range.clone())
+            .unwrap();
+        assert_eq!(result, Some(sealed_headers));
+    }
+
+    #[tokio::test]
+    async fn cached_view__get_sealed_headers__cache_miss() {
+        // given
+        let sender = Arc::new(Notify::new());
+        let db = FakeDb {
+            sender: sender.clone(),
+            values: true,
+        };
+        let cached_view = CachedView::new(false);
+
+        // when
+        let notified = sender.notified();
+        let block_height_range = 0..100;
+        let sealed_headers = default_sealed_headers(block_height_range.clone());
+        let result = cached_view
+            .get_sealed_headers(&db, block_height_range.clone())
```

Comment:
I would expect the cache to be linked to the DB at the time it is created, rather than having to specify the DB when invoking the function `get_sealed_headers` or `get_transactions`. Just curious to know what's the reason behind this choice?

Reply:
you will notice that the view of the current tip of the db (LatestView) is passed into the CachedView while making calls